File: app/src/main/java/com/gystry/pjetpack/ui/home/HomeFragment.java
package com.gystry.pjetpack.ui.home;

import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.Observer;
import androidx.paging.ItemKeyedDataSource;
import androidx.paging.PagedList;
import androidx.paging.PagedListAdapter;

import com.gystry.libnavannotation.FragmentDestination;
import com.gystry.pjetpack.exoplayer.PageListPlayDetector;
import com.gystry.pjetpack.model.Feed;
import com.gystry.pjetpack.ui.AbsListFragment;
import com.gystry.pjetpack.ui.MutablePageKeyedDataSource;
import com.scwang.smartrefresh.layout.api.RefreshLayout;

import java.util.List;

@FragmentDestination(pageUrl = "main/tabs/home", asStarter = true)
public class HomeFragment extends AbsListFragment<Feed, HomeViewModel> {

    private HomeViewModel homeViewModel;
    private PageListPlayDetector pageListPlayDetector;
    private String feedType;
    private boolean shouldPause = true;

    public static HomeFragment newInstance(String feedType) {

        Bundle args = new Bundle();
        args.putString("feedType", feedType);
        HomeFragment fragment = new HomeFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        mViewModel.getCacheLiveData().observe(getViewLifecycleOwner(), new Observer<PagedList<Feed>>() {
            @Override
            public void onChanged(PagedList<Feed> feeds) {
                submitList(feeds);
            }
        });
        pageListPlayDetector = new PageListPlayDetector(this, recyclerView);
        mViewModel.setFeedType(feedType);
    }

    @Override
    public PagedListAdapter getAdapter() {
        feedType = getArguments() == null ? "all" : getArguments().getString("feedType");
        return new FeedAdapter(getContext(), feedType) {
            //检测当列表滑动的时候，内部的item进入和滑出屏幕
            @Override
            public void onViewAttachedToWindow(@NonNull ViewHolder holder) {
                super.onViewAttachedToWindow(holder);
                if (holder.isVideoItem()) {
                    pageListPlayDetector.addTarget(holder.getListPlayerView());
                }
            }

            @Override
            public void onViewDetachedFromWindow(@NonNull ViewHolder holder) {
                super.onViewDetachedFromWindow(holder);
                pageListPlayDetector.removeTarget(holder.getListPlayerView());
            }

            @Override
            public void onStartFeedDetailActivity(Feed feed) {
                boolean isVideo = feed.itemType == Feed.VIDEO_TYPE;
                shouldPause = !isVideo;
            }

            @Override
            public void onCurrentListChanged(@Nullable PagedList<Feed> previousList, @Nullable PagedList<Feed> currentList) {
                //这个方法是在我们每提交一次 pagelist对象到adapter就会触发一次
                //每调用一次adapter.submitlist
                if (previousList != null && currentList != null) {
                    if (!currentList.containsAll(previousList)) {
                        recyclerView.scrollToPosition(0);
                    }
                }
            }
        };
    }

    @Override
    public void onLoadMore(@NonNull RefreshLayout refreshLayout) {
        PagedList<Feed> currentList = adapter.getCurrentList();
        if (currentList == null || currentList.size() <= 0) {
            finishRefresh(false);
            return;
        }
        Feed feed = currentList.get(adapter.getItemCount() - 1);
        mViewModel.loadAfter(feed.id, new ItemKeyedDataSource.LoadCallback<Feed>() {
            @Override
            public void onResult(@NonNull List<Feed> data) {
                PagedList.Config config = currentList.getConfig();
                if (data != null && data.size() > 0) {
                    MutablePageKeyedDataSource dataSource = new MutablePageKeyedDataSource<>();
                    dataSource.data.addAll(currentList);
                    dataSource.data.addAll(data);
                    PagedList pagedList = dataSource.buildNewPagedList(config);
                    submitList(pagedList);
                }
            }
        });
    }

    @Override
    public void onRefresh(@NonNull RefreshLayout refreshLayout) {
        //invalidate 之后Paging会重新创建一个DataSource 重新调用它的loadInitial方法加载初始化数据
        //详情见：LivePagedListBuilder#compute方法
        mViewModel.getDataSource().invalidate();
    }

    @Override
    public void onHiddenChanged(boolean hidden) {
        super.onHiddenChanged(hidden);
        if (hidden) {
            pageListPlayDetector.onPause();
        } else {
            pageListPlayDetector.onResume();
        }
    }

    @Override
    public void onPause() {
        if (shouldPause) {
            pageListPlayDetector.onPause();
        }
        super.onPause();
    }


    @Override
    public void onResume() {
        //解决从后台进入沙发视频页面时，出现首页和沙发视频页面都同时播放的问题
        if (getParentFragment() != null) {
            if (getParentFragment().isVisible()) {
                pageListPlayDetector.onResume();
            }
        } else {
            if (isVisible()) {
                pageListPlayDetector.onResume();
            }
        }
        super.onResume();
    }
}

File: appkt/src/main/java/com/gystry/appkt/ui/home/HomeFragment.kt
package com.gystry.appkt.ui.home

import androidx.lifecycle.Observer
import androidx.lifecycle.observe
import androidx.paging.ItemKeyedDataSource
import androidx.paging.PagedList
import androidx.paging.PagedListAdapter
import com.gystry.appkt.model.Feed
import com.gystry.appkt.ui.AbsListFragment
import com.gystry.appkt.ui.MutablePageKeyedDataSource
import com.gystry.libnavannotation.FragmentDestination
import com.scwang.smartrefresh.layout.api.RefreshLayout

@FragmentDestination(pageUrl = "main/tabs/home", asStarter = true)
class HomeFragment : AbsListFragment<Feed, HomeViewModel, FeedAdapter.ViewHolder>() {

    override fun getAdapter(): PagedListAdapter<Feed, FeedAdapter.ViewHolder> {
        val feedType = if (arguments == null) "all" else requireArguments().getString("fedType")
        return FeedAdapter(context, feedType!!)
    }
    override fun onLoadMore(refreshLayout: RefreshLayout) {
        val feed = getAdapter().currentList?.get(getAdapter().itemCount - 1)
        mViewModel.loadAfter(feed?.id!!,object :ItemKeyedDataSource.LoadCallback<Feed>(){
            override fun onResult(data: MutableList<Feed>) {
                val config = getAdapter().currentList?.config
                if (data!=null&&data.size>=0) {
            val dataSource=MutablePageKeyedDataSource<Feed>()
                    dataSource.data.addAll(data)
                    val buildNewPagedList = dataSource.buildNewPagedList(config!!)
                    mViewModel.cacheLiveData.postValue(buildNewPagedList)
                }
            }

        })
    }

    override fun onRefresh(refreshLayout: RefreshLayout) {
        //当下拉刷新的时候，调用datasource的invalidate
        mViewModel.getDataSource()?.invalidate()
    }

    override fun afterCreateView() {
        mViewModel.cacheLiveData.observe(this){
            getAdapter().submitList(it)
        }
    }

}
//提交一次
//提交two次
//提交3次